archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+    left_inv := by { intro f, simp },
+    right_inv := by { intro f, simp }}},
+
+  refine subtype_equiv (perm_splice_out a b) _,
+  rintro ⟨f, hf⟩,
+  exact splice_derangement a b hab' f hf,
+end
+
+/-- For a given `a b : α`, the set of derangements sending `a` to `b` is equivalent to
+    `derangements {a,b}ᶜ ⊕ derangements {a}ᶜ`. -/
+def equiv_one_fiber (a : α) (b : everything_but a) :
+  {f : derangements α // f a = b} ≃
+  (derangements (neither_of a b) ⊕ derangements (everything_but a)) :=
+begin
+  -- split the type into those that send b back to a, and those that don't
+  let b_goes_back_to_a : derangements α → Prop := λ f, f b = a,
+
+  refine equiv.trans (sum_compl (b_goes_back_to_a ∘ coe)).symm _,
+  apply equiv.sum_congr,
+  { exact (subtype_subtype_equiv_subtype_inter _ _).trans (equiv_of_ignore_swaps a b) },
+  { calc {f : {f : derangements α // f a = b} // ¬(b_goes_back_to_a ∘ coe) f }
+        ≃ {f : derangements α // f a = b ∧ ¬b_goes_back_to_a f}
+            : subtype_subtype_equiv_subtype_inter _ (not ∘ b_goes_back_to_a)
+    ... ≃ derangements (everything_but a)
+            : equiv_of_splice_out a b }
+end
+
+/-- For any fixed `a : α`, the set of derangements of `α` is equivalent to the disjoint union of
+    `derangements {a,b}ᶜ ⊕ derangements {a}ᶜ`) over all `b ≠ a`. -/
+def derangements_recursion_equiv (a : α):
+  derangements α ≃
+  Σ b : everything_but a, (derangements (neither_of a b) ⊕ derangements (everything_but a)) :=
+begin
+  -- the fiber over b = a is empty
+  have fact : ∀ (b : α), {f : derangements α // f a = b} → b ∈ everything_but a,
+  { rintros b ⟨⟨f, f_derangement⟩, fa_eq_b⟩,
+    rw ← fa_eq_b,
+    exact f_derangement a },
+
+  calc derangements α
+      ≃ Σ b : α, {f : derangements α // f a = b}
+          : (sigma_preimage_equiv _).symm
+  ... ≃ Σ b : everything_but a, {f : derangements α // f a = b}
+          : (sigma_subtype_equiv_of_subset _ _ fact).symm
+  ... ≃ Σ b : everything_but a, (derangements (neither_of a b) ⊕ derangements (everything_but a))
+          : sigma_congr_right (equiv_one_fiber a),
+end
+
+end recursion
+
+
+section finite_instances
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (@is_derangement α) :=
+begin
+  intro f,
+  apply fintype.decidable_forall_fintype,
+end
+
+instance : fintype (derangements α) :=
+begin
+  have : fintype (perm α) := by apply_instance,
+  dsimp [derangements],
+  exact set_fintype (set_of is_derangement),
+end
+
+instance (a : α) : fintype (everything_but a) :=
+begin
+  simp [everything_but],
+  apply fintype.of_finset ({a}ᶜ : finset α),
+  simp,
+end
+
+instance (a : α) (b : everything_but a) : fintype (neither_of a b) :=
+begin
+  simp [everything_but, neither_of],
+  apply fintype.of_finset ({a, b}ᶜ : finset α),
+  simp,
+end
+
+end finite_instances
+
+
+section counting
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
```

Comment:
I'm not sure it's even worth introducing this. How does it look to just write it out each time?

Reply:
IMO it's like the catalan numbers, where you can just say `a(n)` without reference to a specific set of n elements

not super necessary but i think it's nice